src/elf_sections.rs

```diff
@@ -18,6 +18,40 @@ impl ElfSectionsTag {
             entry_size: self.entry_size,
         }
     }
+
+    pub fn string_table(&self) -> &'static StringTable {
+        unsafe {
+            let string_table_ptr =
+                (&self.first_section as *const ElfSection).offset(self.shndx as isize);
+            &*((*string_table_ptr).addr as *const StringTable)
+        }
+    }
+}
+
+pub struct StringTable(u8);
+
+impl StringTable {
+    pub fn section_name(&self, section: &ElfSection) -> &'static str {
+        unsafe fn strlen(start: *const u8) -> usize {
```

Comment:
Only `*start.offset(i)` is unsafe, so we could just do:

```rust
for i in 0.. {
    if unsafe { *start.offset(i) } == 0 {
        return i as usize;
    }
}
unreachable!()
```

This way it's much easier to spot the unsafe operations. On the other hand, I'm not comfortable with removing the `unsafe` from the `strlen` function (since it's not safe).

So I would prefer to remove the `strlen` function. Instead, we could do it like in your first commit:

```rust
let strlen;
for i in 0.. {
    if unsafe { *start.offset(i) } == 0 {
        strlen = i as usize;
        break;
    }
};
```

Or alternatively with a `while` loop:

```rust
let strlen = {
    let mut len = 0;
    while unsafe { *start.offset(len) } != 0 {
        len += 1;
    }
    len as usize
};
```